### PR TITLE
EASY-1507: Add Base Revision to instructions.csv

### PR DESCRIPTION
**Pull request by gulcinermis**
fixes EASY-1507
* [x] done

#### When applied it will
* add the uuid of the base pointed by the current deposit to bag-info
* validate base revision conforms to UUID pattern 
* (used an online UUID generator for UUID s in instructions file)

#### related pull requests on github
repo                       | PR
-------------------------- | -----------------
easy-                      | [PR#](PRlink)

### CODE REVIEW

**Comment by gulcinermis** [comment]
@rvanheest I haven't implemented all your requests in your e-mail yet but in this version of the project the base uuid is successfully recorded in bag-info. I realized that I wasn't entering the .combine(..) list in the right order before. I also added a validation rule to validatePreconditions and the related spec, however, although the rule works as expected in mvnci, it doesn't give a failure on VM, I will look at it later. Also I will add the other unit tests regarding the baseUUID.

File: src/main/scala/nl.knaw.dans.easy.multideposit/SplitMultiDepositApp.scala [hunk 1]
```diff
@@ -77,6 +77,7 @@ class SplitMultiDepositApp(formats: Set[String], ldap: Ldap, permissions: Deposi
 
   private def convertDeposit(paths: PathExplorers, datamanagerId: Datamanager, dataManagerEmailAddress: DatamanagerEmailaddress)(deposit: Deposit): Try[Unit] = {
     val depositId = deposit.depositId
+    val base = deposit.baseUUID
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/SplitMultiDepositApp.scala]: inline, since only one use

File: src/main/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDeposit.scala [hunk 2]
```diff
@@ -25,7 +25,10 @@ import gov.loc.repository.bagit.verify.FileCountAndTotalSizeVistor
 import nl.knaw.dans.easy.multideposit.FileExtensions
 import nl.knaw.dans.easy.multideposit.PathExplorer.{ InputPathExplorer, StagingPathExplorer }
 import nl.knaw.dans.easy.multideposit.model.DepositId
+import nl.knaw.dans.easy.multideposit.parser.MultiDepositParser.parse
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDeposit.scala]: unused imports

File: src/main/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDeposit.scala [hunk 3]
```diff
@@ -34,25 +37,28 @@ import scala.util.{ Failure, Try }
...
     val metadata = new BagitMetadata {
       add("Created", created.toString(ISODateTimeFormat.dateTime()))
     }
 
+    metadata.add("Is-Version-Of", base )
+
     if (Files.exists(inputDir)) {
       inputDir.copyDir(stageDir)
       metadata.add("Bag-Size", formatSize(calculateSizeOfPath(inputDir)))
+
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDeposit.scala]: unnecessary newlines here and on lines 102 and 103

File: src/main/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditions.scala [hunk 4]
```diff
@@ -115,4 +117,13 @@ class ValidatePreconditions(ldap: Ldap) extends DebugEnhancedLogging {
         case false => Failure(InvalidInputException(deposit.row, s"The depositor '$depositorUserId' is not an active user"))
       }
   }
+
+  //TODO This function works as intended when mvnci applied however I did not get any failure message on VM with an invalid UUID
+  def checkBaseRevisionConformsToUUID(deposit:Deposit, base:String): Try[Unit] = {
+    val depositorBaseRevision = base
+    Try(UUID.fromString(depositorBaseRevision)).isFailure match {
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditions.scala]: nice check! Didn't think of that when we discussed the issue. Perhaps tidy it up a bit with

```scala
Try(UUID.fromString(...)).recoverWith {
  case e => Failure(InvalidInputException(deposit.row, "base revision is not in UUID format"))
}
```

Besides, please run the formatter on this function.

File: src/main/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditions.scala [hunk 5]
```diff
@@ -16,10 +16,12 @@
 package nl.knaw.dans.easy.multideposit.actions
 
 import java.nio.file.{ Files, Path }
+import java.util.UUID
 
 import nl.knaw.dans.easy.multideposit.Ldap
 import nl.knaw.dans.easy.multideposit.PathExplorer.{ OutputPathExplorer, StagingPathExplorer }
 import nl.knaw.dans.easy.multideposit.model.{ AVFileMetadata, Deposit, DepositId }
+import nl.knaw.dans.easy.multideposit.parser.MultiDepositParser
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditions.scala]: unused import?

File: src/main/scala/nl.knaw.dans.easy.multideposit/model/package.scala [hunk 6]
```diff
@@ -25,6 +25,7 @@ package object model {
   type MimeType = String
   type Datamanager = String
   type DatamanagerEmailaddress = String
+  type BaseUUID = String
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/model/package.scala]: Since it's a UUID, why don't you make it a `UUID` type?

File: src/main/scala/nl.knaw.dans.easy.multideposit/parser/Headers.scala [hunk 7]
```diff
@@ -15,7 +15,10 @@
  */
 package nl.knaw.dans.easy.multideposit.parser
 
-object Headers {
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+import nl.knaw.dans.lib.error._
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/parser/Headers.scala]: unused import

File: src/main/scala/nl.knaw.dans.easy.multideposit/parser/Headers.scala [hunk 8]
```diff
@@ -34,6 +37,8 @@ object Headers {
       audioVideoHeaders
   }
 
+  logger.info(s"VALID HEADERS $validHeaders")
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/parser/Headers.scala]: 👍 I like this addition.
However, make it `debug` rather than `info`. With `info`, the datamanager will see this log everytime the tool is used. That's not needed, I think.

File: src/main/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParser.scala [hunk 9]
```diff
@@ -112,7 +113,7 @@ trait MultiDepositParser extends ParserUtils with InputPathExplorer
       .map(_ => ())
   }
 
-  def extractDeposit(multiDepositDirectory: Path)(depositId: DepositId, rows: DepositRows): Try[Deposit] = {
+  def extractDeposit(multiDepositDirectory: Path)(depositId: DepositId, rows: DepositRows ): Try[Deposit] = {
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParser.scala]: ??? why this change?

File: src/main/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParser.scala [hunk 10]
```diff
@@ -122,13 +123,14 @@ trait MultiDepositParser extends ParserUtils with InputPathExplorer
...
     val rowNum = rows.map(getRowNum).min
-
     checkValidChars(depositId, rowNum, "DATASET")
       .flatMap(dsId => Try { Instructions.curried }
         .map(_ (dsId))
         .map(_ (rowNum))
         .combine(extractNEL(rows, rowNum, "DEPOSITOR_ID").flatMap(exactlyOne(rowNum, List("DEPOSITOR_ID"))))
+        //.combine (atMostOne(rowNum, List("BASE_REVISION"))(extractList(rows, "BASE_REVISION").map(string => UUID.fromString(string))))
         .combine(extractProfile(rows, rowNum))
+        .combine(Try(atMostOne(rowNum, List("BASE_REVISION"))(extractList(rows, "BASE_REVISION")).get.get))
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParser.scala]: we need to look at this line next week

1. The `.get`s don't seem right to me
2. `checkBaseRevisionConformsToUUID` needs to be incorporated here

File: src/main/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParser.scala [hunk 11]
```diff
@@ -163,6 +165,7 @@ trait MultiDepositParser extends ParserUtils with InputPathExplorer
   }
 }
 
+
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParser.scala]: newline doesn't seem necessary

File: src/test/resources/invalidCSV/input/instructions.csv [hunk 12]
```diff
@@ -1,4 +1,4 @@
...
+ruimtereis01,Reis naar Centaur-planetoïde,Gewoon een stukje steen in de ruimte?,Captain,J.T.,,Kirk,,United Federation of Planets,Prof. Dr. Mr. Ir.,A.S.,,Terix,info:eu-repo/dai/nl/123456789X,Gallië,invalid-date,jan,dans,Een tweedaagse reis per ruimteschip naar een bijzondere planetoïde in de omgeving van Jupiter.,astronomie,,PALEOLB,abr:ABRperiode,here,degrees,83575.4,455271.2,,,,,ds1,,replaces,dummy,,Text with Qualifier,ISSUED,random test data,video/mpeg,NL,whatever,OPEN_ACCESS,1992-07-30,,path/to/audiofile/that/does/not/exist.mp3,,,other#domain,,other collection,,,,,user001,ab23-bdcg
```
Comment (rvanheest) [inline src/test/resources/invalidCSV/input/instructions.csv]: great to see you add an invalid UUID here! Hadn't even thought about this, but this way you can immediately test your new validation rule (`checkBaseRevisionConformsToUUID`)!

File: src/test/scala/nl.knaw.dans.easy.multideposit/TestSupportFixture.scala [hunk 13]
```diff
@@ -93,6 +94,7 @@ trait TestSupportFixture extends FlatSpec with Matchers with OptionValues with I
         audiences = List("D37000"),
         accessright = AccessCategory.GROUP_ACCESS
       ),
+      baseUUID = "ab23-bdgg",
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.multideposit/TestSupportFixture.scala]: this is not a valid UUID and this example is supposed to be valid

File: src/test/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDepositSpec.scala [hunk 14]
```diff
@@ -31,6 +31,7 @@ class AddBagToDepositSpec extends TestSupportFixture with BeforeAndAfterEach {
 
   private val depositId = "dsId1"
   private val date = new DateTime(1992, 7, 30, 0, 0)
+  private val base = "ab23-bdcg"
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDepositSpec.scala]: please use valid UUID's in these circumstances (not only here, but also in other test files). That's why you should use the `UUID` type instead. That one guarantees that automatically for you.

File: src/test/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditionsSpec.scala [hunk 15]
```diff
@@ -215,4 +216,16 @@ class ValidatePreconditionsSpec extends TestSupportFixture with BeforeAndAfterEa
       case Failure(ActionException(message, _)) => message should include("multiple users with id 'dp1'")
     }
   }
+
+  "checkBaseRevisionConformsToUUID" should "not fail if the base revision conforms to UUID pattern" in {
+    val base = testInstructions1.copy(depositorUserId = "dp1").toDeposit().baseUUID
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditionsSpec.scala]: extract the common bit from these lines

File: src/test/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditionsSpec.scala [hunk 16]
```diff
@@ -215,4 +216,16 @@ class ValidatePreconditionsSpec extends TestSupportFixture with BeforeAndAfterEa
...
+    val row = testInstructions1.copy(depositorUserId = "dp1").toDeposit().row
+    action.checkBaseRevisionConformsToUUID(testInstructions1.copy(depositorUserId = "dp1").toDeposit(), base) shouldBe Success(())
+  }
+
+  it should "fail if the base revision does not conform to UUID pattern" in {
+    val base = testInstructions2.copy(depositorUserId = "dp1").toDeposit().baseUUID
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditionsSpec.scala]: I now see why you made `testInstructions2` invalid. However, I think it is better to leave `testInstructions2` valid, as it was before. Then just use `.copy` to create an invalid one.

File: src/test/resources/invalidCSV/input/instructions.csv (hunk 12 above)
Comment (rvanheest) [reply src/test/resources/invalidCSV/input/instructions.csv]: after further reviewing, I would expect this error to show up in the test suite. However, I don't see it yet. I know you're not finished, but please don't forget to add the check.

**Review by rvanheest — CHANGES_REQUESTED** [review]
Looks nice already. There are a couple of points you need to further look into. Besides, please have a look at the complete list of steps that I send you earlier this week via email. I'm still missing some points in this PR.

File: src/test/scala/nl.knaw.dans.easy.multideposit/TestSupportFixture.scala (hunk 13 above)
Comment (gulcinermis) [reply src/test/scala/nl.knaw.dans.easy.multideposit/TestSupportFixture.scala]: replaced it with a valid one an created a new testInstructions with invalid uuid

**Comment by gulcinermis** [comment]
@rvanheest I was about to add a new check to test if all "uuid"s in instructions file for a deposit are the same or not (in the event they exist) regarding your e-mail but the excessive uuid data for one deposit is already forbidden in the existing code, thus there is no need to add this check.
Example:
row 2: Only one row is allowed to contain a value for the column 'BASE_REVISION'. Found: [d5e8f0fb-c374-86eb-918c-b06dd5ae5e71, d5e8f0fb-c372-86eb-918c-b06dd5ae5e71]

File: src/test/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParserSpec.scala [hunk 17]
```diff
@@ -331,4 +331,13 @@ class MultiDepositParserSpec extends TestSupportFixture with DepositTestObjects
       case Success(_) =>
     }
   }
+
+  it should "fail if there are multiple distinct base revisions" in {
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParserSpec.scala]: nice test! However, what you're testing is `extractInstructions`, not `uuid`. The `it` at the start of this line is in the context of `uuid`, so it looks like this is part of the test suite for the `uuid` function. However, it is rather meant to be part of the test suite of `extractInstructions`. Please move the test to the correct position.

File: src/test/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParserSpec.scala [hunk 18]
```diff
@@ -331,4 +331,13 @@ class MultiDepositParserSpec extends TestSupportFixture with DepositTestObjects
...
+    val ROW = Map("ROW" -> "3", "DATASET" -> "ruimtereis01") ++ profileCSVRow2 ++ Map("BASE_REVISION" -> "9de3f841-0f0d-048b-b3db-4b03ad4834d7") ++ metadataCSVRow2 ++ fileDescriptorCSVRow2 ++ audioVideoCSVRow2
+    val rows = depositCSVRow1 :: ROW :: Nil
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParserSpec.scala]: after moving... can you minimize these two lines by looking at the other examples in this test suite?

File: src/test/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParserSpec.scala (hunk 17 above)
Comment (rvanheest) [reply src/test/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParserSpec.scala]: Oeps... You already did that 😜

File: src/test/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDepositSpec.scala [hunk 19]
```diff
@@ -129,6 +129,26 @@ class AddBagToDepositSpec extends TestSupportFixture with BeforeAndAfterEach {
...
+    action.addBagToDeposit(depositId, date, base) shouldBe a[Success[_]]
+
+    val bagInfo = stagingBagDir(depositId).resolve("bag-info.txt")
+    bagInfo.toFile should exist
+
+    bagInfo.read() should include("Is-Version-Of: " + base.get)
+  }
+
+  it should "not contain the Is-Version-Of in the bag-info.txt if the Option[BaseUUID] is None " in {
+    base = None
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDepositSpec.scala]: no need to make `base` a variable rather than a value. Just plug in `None` to `action.addBagToDeposit(depositId, date, None)` instead.

File: src/test/resources/allfields/output/input-ruimtereis01/bag/bag-info.txt [hunk 20]
```diff
@@ -1,4 +1,5 @@
 Payload-Oxum: 25300962.9
+Is-Version-Of: d5e8f0fb-c374-86eb-918c-b06dd5ae5e71
```
Comment (rvanheest) [inline src/test/resources/allfields/output/input-ruimtereis01/bag/bag-info.txt:2]: I would guess that the checksum (`sha1`) for these resulting bags should change as well in `tagmanifest-sha1.txt`? If the test doesn't fail due to that, probably I made a mistake a while back 😄 
Please let me know if this is the case!

File: multi-deposit-instructions.md [hunk 21]
```diff
@@ -260,3 +260,13 @@ combined with the `DATASET` value.
 [refinements of the relation element]: http://dublincore.org/documents/usageguide/qualifiers.shtml#isVersionOf
 [Springfield Web TV]: http://noterik.github.io/
 [DataCite ContributorType list]: http://schema.datacite.org/meta/kernel-4.0/include/datacite-contributorType-v4.xsd
+
+2. BASE_REVISION column
+-----------------------
+
+UUID values of the base bags pointed by the deposits are stored in this column.
```
Comment (rvanheest) [inline multi-deposit-instructions.md]: ...pointed **_to_** by...

File: multi-deposit-instructions.md [hunk 22]
```diff
@@ -260,3 +260,13 @@ combined with the `DATASET` value.
...
+
+In the instructions.csv file, the list of uuid values in the `BASE_REVISON` column must have the same value
```
Comment (rvanheest) [inline multi-deposit-instructions.md]: _instructions.csv_ as inline code snippet using **_\`instructions.csv\`_**

File: multi-deposit-instructions.md [hunk 23]
```diff
@@ -260,3 +260,13 @@ combined with the `DATASET` value.
...
 [Springfield Web TV]: http://noterik.github.io/
 [DataCite ContributorType list]: http://schema.datacite.org/meta/kernel-4.0/include/datacite-contributorType-v4.xsd
+
+2. BASE_REVISION column
+-----------------------
+
+UUID values of the base bags pointed by the deposits are stored in this column.
+
+In the instructions.csv file, the list of uuid values in the `BASE_REVISON` column must have the same value 
+if they belong to the same dataset, or only one uuid should be provided for one dataset which has to be located
```
Comment (rvanheest) [inline multi-deposit-instructions.md]: ~~dataset~~ deposit

File: multi-deposit-instructions.md [hunk 24]
```diff
@@ -260,3 +260,13 @@ combined with the `DATASET` value.
...
 [DataCite ContributorType list]: http://schema.datacite.org/meta/kernel-4.0/include/datacite-contributorType-v4.xsd
+
+2. BASE_REVISION column
+-----------------------
+
+UUID values of the base bags pointed by the deposits are stored in this column.
+
+In the instructions.csv file, the list of uuid values in the `BASE_REVISON` column must have the same value 
+if they belong to the same dataset, or only one uuid should be provided for one dataset which has to be located 
+at the first row each time a new dataset info is added.
```
Comment (rvanheest) [inline multi-deposit-instructions.md]: not sure about this first row thing. Can you try that (only add a uuid to the second row)?

File: multi-deposit-instructions.md [hunk 25]
```diff
@@ -260,3 +260,13 @@ combined with the `DATASET` value.
 [refinements of the relation element]: http://dublincore.org/documents/usageguide/qualifiers.shtml#isVersionOf
 [Springfield Web TV]: http://noterik.github.io/
 [DataCite ContributorType list]: http://schema.datacite.org/meta/kernel-4.0/include/datacite-contributorType-v4.xsd
+
+2. BASE_REVISION column
```
Comment (rvanheest) [inline multi-deposit-instructions.md]: - I would expect this section to go somewhere in the section `### Semantics of the columns` section (starting at line 74).
- Also it should be somewhere in the listing `Metadata elements` (starting at line 41).

**Review by rvanheest — CHANGES_REQUESTED** [review]

File: src/test/resources/allfields/output/input-ruimtereis01/bag/bag-info.txt (hunk 20 above)
Comment (gulcinermis) [reply src/test/resources/allfields/output/input-ruimtereis01/bag/bag-info.txt:2]: Yes you are right, tests didn't fail :)

File: multi-deposit-instructions.md (hunk 24 above)
Comment (gulcinermis) [reply multi-deposit-instructions.md]: Yes it works for the second row too, I will update the info.

File: src/main/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDeposit.scala (hunk 3 above)
Comment (gulcinermis) [reply src/main/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDeposit.scala]: deleted them but an extra line is always created at the end of the code each time I compile

File: src/test/resources/allfields/output/input-ruimtereis01/bag/bag-info.txt (hunk 20 above)
Comment (rvanheest) [reply src/test/resources/allfields/output/input-ruimtereis01/bag/bag-info.txt:2]: Surprise! That's correct; the test shouldn't fail. See the comment in the following code snippet... Please change the `tagmanifest-sha1.txt` files manually.

https://github.com/DANS-KNAW/easy-split-multi-deposit/blob/1f81280cfb6bacfb43ad34aff7ab8086f656f374/src/test/scala/nl.knaw.dans.easy.multideposit/SplitMultiDepositAppSpec.scala#L179-L184

File: src/test/resources/allfields/input/instructions.csv [hunk 26]
```diff
@@ -1,11 +1,11 @@
...
+DATASET,DC_TITLE,DCT_ALTERNATIVE,DCX_CREATOR_TITLES,DCX_CREATOR_INITIALS,DCX_CREATOR_INSERTIONS,DCX_CREATOR_SURNAME,DCX_CREATOR_DAI,DCX_CREATOR_ORGANIZATION,DCX_CREATOR_ROLE,DCX_CONTRIBUTOR_TITLES,DCX_CONTRIBUTOR_INITIALS,DCX_CONTRIBUTOR_INSERTIONS,DCX_CONTRIBUTOR_SURNAME,DCX_CONTRIBUTOR_DAI,DCX_CONTRIBUTOR_ORGANIZATION,DCX_CONTRIBUTOR_ROLE,DDM_CREATED,DCT_RIGHTSHOLDER,DC_PUBLISHER,DC_DESCRIPTION,DC_SUBJECT,DC_SUBJECT_SCHEME,DCT_TEMPORAL,DCT_TEMPORAL_SCHEME,DCT_SPATIAL,DCX_SPATIAL_SCHEME,DCX_SPATIAL_X,DCX_SPATIAL_Y,DCX_SPATIAL_NORTH,DCX_SPATIAL_SOUTH,DCX_SPATIAL_EAST,DCX_SPATIAL_WEST,DC_IDENTIFIER,DC_IDENTIFIER_TYPE,DCX_RELATION_QUALIFIER,DCX_RELATION_TITLE,DCX_RELATION_LINK,DCT_DATE,DCT_DATE_QUALIFIER,DC_TYPE,DC_FORMAT,DC_LANGUAGE,DC_SOURCE,DDM_ACCESSRIGHTS,DDM_AVAILABLE,DDM_AUDIENCE,FILE_PATH,FILE_TITLE,FILE_ACCESSIBILITY,SF_DOMAIN,SF_USER,SF_COLLECTION,SF_PLAY_MODE,AV_FILE_PATH,AV_SUBTITLES,AV_SUBTITLES_LANGUAGE,DEPOSITOR_ID,BASE_REVISION
```
Comment (rvanheest) [inline src/test/resources/allfields/input/instructions.csv]: 1. let's only put base revisions in `deposit1` and `deposit4`; then we also have test cases without a base revision (which is also still valid, of course)
2. in `deposit1` use the same base revision in multiple rows
2a. change the output's (`bag-info.txt`) accordingly
3. change the `tagmanifest-sha1.txt` files of deposit1 and deposit4 accordingly
4. sync the CSV with the Excel sheet

**Review by rvanheest — COMMENTED** [review]
here are some comments already. I'll start a full review shortly...

File: src/main/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDeposit.scala [hunk 27]
```diff
@@ -24,8 +24,11 @@ import gov.loc.repository.bagit.hash.StandardSupportedAlgorithms
 import gov.loc.repository.bagit.verify.FileCountAndTotalSizeVistor
 import nl.knaw.dans.easy.multideposit.FileExtensions
 import nl.knaw.dans.easy.multideposit.PathExplorer.{ InputPathExplorer, StagingPathExplorer }
-import nl.knaw.dans.easy.multideposit.model.DepositId
+import nl.knaw.dans.easy.multideposit.model.{ BaseUUID, DepositId }
+import nl.knaw.dans.easy.multideposit.parser.MultiDepositParser.parse
+import nl.knaw.dans.easy.multideposit.parser.{ DepositRow, DepositRows, MultiDepositParser, ParserUtils }
 import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+import org.apache.commons.csv.CSVParser
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDeposit.scala]: I think there are some unused imports here...

File: src/main/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditions.scala [hunk 28]
```diff
@@ -36,6 +37,7 @@ class ValidatePreconditions(ldap: Ldap) extends DebugEnhancedLogging {
       _ <- checkSFColumnsIfDepositContainsAVFiles(deposit)
       _ <- checkEitherVideoOrAudio(deposit)
       _ <- checkDepositorUserId(deposit)
+      //_ <- checkBaseRevisionConformsToUUID(deposit)
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditions.scala]: Please remove all changes in this class. It turned out you didn't need any additions in here.

File: src/main/scala/nl.knaw.dans.easy.multideposit/parser/Headers.scala (hunk 7 above)
Comment (rvanheest) [reply src/main/scala/nl.knaw.dans.easy.multideposit/parser/Headers.scala]: friendly reminder 😉

File: src/main/scala/nl.knaw.dans.easy.multideposit/parser/ProfileParser.scala [hunk 29]
```diff
@@ -17,10 +17,10 @@ package nl.knaw.dans.easy.multideposit.parser
 
 import nl.knaw.dans.common.lang.dataset.AccessCategory
 import nl.knaw.dans.easy.multideposit.model._
+import nl.knaw.dans.lib.error._
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.multideposit/parser/ProfileParser.scala]: unused import

File: src/test/scala/nl.knaw.dans.easy.multideposit/SplitMultiDepositAppSpec.scala [hunk 30]
```diff
@@ -100,6 +100,9 @@ class SplitMultiDepositAppSpec extends TestSupportFixture with MockFactory with
 
     it should "succeed validating the multideposit" in {
       configureLdapMockBehavior()
+      //val base = Deposit.
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.multideposit/SplitMultiDepositAppSpec.scala]: please remove these lines

File: src/test/scala/nl.knaw.dans.easy.multideposit/SplitMultiDepositAppSpec.scala [hunk 31]
```diff
@@ -145,6 +148,8 @@ class SplitMultiDepositAppSpec extends TestSupportFixture with MockFactory with
         // skipping the Bagging-Date which is different every time
         bagInfo.read().lines.toSeq should contain allElementsOf
           expBagInfo.read().lines.filterNot(_ contains "Bagging-Date").toSeq
+
+        bagInfo.read().lines.filterNot(_ contains "Bagging-Date").toSet should contain theSameElementsAs expBagInfo.read().lines.filterNot(_ contains "Bagging-Date").toSet
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.multideposit/SplitMultiDepositAppSpec.scala]: I meant to replace the test above (line 149-150) with this test, rather than adding it

File: src/test/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditionsSpec.scala [hunk 32]
```diff
@@ -28,6 +28,7 @@ import scala.util.{ Failure, Success, Try }
 class ValidatePreconditionsSpec extends TestSupportFixture with BeforeAndAfterEach with MockFactory {
 
   private val depositId = "dsId1"
+  //private val depositorBaseRevision = "xx-dd"
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.multideposit/actions/ValidatePreconditionsSpec.scala]: please remove the commented out lines here and lines 219-238

File: src/test/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParserSpec.scala [hunk 33]
```diff
@@ -33,15 +34,16 @@ trait DepositTestObjects extends AudioVideoTestObjects
   this: TestSupportFixture =>
 
   lazy val depositCSV @ depositCSVRow1 :: depositCSVRow2 :: Nil = List(
-    Map("ROW" -> "2", "DATASET" -> "ruimtereis01", "DEPOSITOR_ID" -> "ikke") ++ profileCSVRow1 ++ metadataCSVRow1 ++ fileDescriptorCSVRow1 ++ audioVideoCSVRow1,
-    Map("ROW" -> "3", "DATASET" -> "ruimtereis01") ++ profileCSVRow2 ++ metadataCSVRow2 ++ fileDescriptorCSVRow2 ++ audioVideoCSVRow2
+    Map("ROW" -> "2", "DATASET" -> "ruimtereis01", "DEPOSITOR_ID" -> "ikke") ++ profileCSVRow1 ++ Map("BASE_REVISION" -> "1de3f841-0f0d-048b-b3db-4b03ad4834d7") ++ metadataCSVRow1 ++ fileDescriptorCSVRow1 ++ audioVideoCSVRow1,
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.multideposit/parser/MultiDepositParserSpec.scala]: just add the base revision to the `Map` at the start of this line instead. No need to create a new `Map`. Same for the line below.

```scala
Map("ROW" -> "2", "DATASET" -> "ruimtereis01", "DEPOSITOR_ID" -> "ikke", "BASE_REVISION" -> "1de3f841-0f0d-048b-b3db-4b03ad4834d7") ++ ...
```

**Review by rvanheest — CHANGES_REQUESTED** [review]
here are some more comments, mainly cleanup of 'old code'

**Comment by rvanheest** [comment]
I forgot to mention one thing in my review: add an invalid base revision to the invalidCSV example (https://github.com/DANS-KNAW/easy-split-multi-deposit/tree/master/src/test/resources/invalidCSV) and change the failing test accordingly. Also keep the corresponding Excel file and README in sync.

File: src/main/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDeposit.scala (hunk 3 above)
Comment (rvanheest) [reply src/main/scala/nl.knaw.dans.easy.multideposit/actions/AddBagToDeposit.scala]: That's right. In IntelliJ you should see this empty line at the end of each file.

File: src/test/resources/allfields/input/instructions.csv [hunk 34]
```diff
@@ -1,11 +1,11 @@
...
...URNAME,DCX_CONTRIBUTOR_DAI           ,DCX_CONTRIBUTOR_ORGANIZATION,DCX_CONTRIBUTOR_ROLE,DDM_CREATED,DCT_RIGHTSHOLDER,DC_PUBLISHER,DC_DESCRIPTION                                                                                                                                                                                                                                                                                                                   ,DC_SUBJECT ,DC_SUBJECT_SCHEME,DCT_TEMPORAL,DCT_TEMPORAL_SCHEME,DCT_SPATIAL,DCX_SPATIAL_SCHEME,DCX_SPATIAL_X,DCX_SPATIAL_Y,DCX_SPATIAL_NORTH,DCX_SPATIAL_SOUTH,DCX_SPATIAL_EAST,DCX_SPATIAL_WEST,DC_IDENTIFIER,DC_IDENTIFIER_TYPE       ,DCX_RELATION_QUALIFIER,DCX_RELATION_TITLE,DCX_RELATION_LINK,DCT_DATE   ,DCT_DATE_QUALIFIER,DC_TYPE,DC_FORMAT                                                              ,DC_LANGUAGE,DC_SOURCE                ,DDM_ACCESSRIGHTS                ,DDM_AVAILABLE,DDM_AUDIENCE,FILE_PATH                         ,FILE_TITLE                       ,FILE_ACCESSIBILITY,SF_DOMAIN,SF_USER     ,SF_COLLECTION  ,SF_PLAY_MODE,AV_FILE_PATH           ,AV_SUBTITLES                      ,AV_SUBTITLES_LANGUAGE,DEPOSITOR_ID,BASE_REVISION
```
Comment (rvanheest) [inline src/test/resources/allfields/input/instructions.csv]: Does this actually work, with this extra formatting? That would be great for future development!

Comment (rvanheest) [reply src/test/resources/allfields/input/instructions.csv]: In that case, also format the invalid `instructions.csv`

Comment (gulcinermis) [reply src/test/resources/allfields/input/instructions.csv]: It really works :)

Comment (rvanheest) [reply src/test/resources/allfields/input/instructions.csv]: awesome! 🎉

File: src/main/scala/nl.knaw.dans.easy.multideposit/parser/ProfileParser.scala (hunk 29 above)
Comment (gulcinermis) [reply src/main/scala/nl.knaw.dans.easy.multideposit/parser/ProfileParser.scala]: when I remove import nl.knaw.dans.lib.error._ ,
'.combine(extract..........)' becomes red.
optimize imports also doesn't remove it, 
maybe we made some changes in extract functions

Comment (rvanheest) [reply src/main/scala/nl.knaw.dans.easy.multideposit/parser/ProfileParser.scala]: I'm sorry, I thought this was just an unused import. What I didn't see at first in the diff on GitHub was that this import was actually moved (probably by doing a format) to it's correct place. It's good as it is right now.

**Comment by gulcinermis** [comment]
I use val base = Option(........) in the rest of the functions, thus I left the base variable as option. 
For the function including base.get, I replaced the base.get by the ..... value from Option(...) to avoid codacy warning.
